precompile/allow_list.go

```diff
@@ -0,0 +1,207 @@
+type AllowListConfig struct {
+	BlockTimestamp *big.Int `json:"blockTimestamp"`
+
+	AllowListAdmins []common.Address `json:"adminAddresses"`
+}
+
+// Address returns the address where the stateful precompile is accessible.
+func (c *AllowListConfig) Address() common.Address { return AllowListAddress }
+
+// Timestamp returns the timestamp at which the allow list should be enabled
+func (c *AllowListConfig) Timestamp() *big.Int { return c.BlockTimestamp }
+
+// Configure initializes the address space of [ModifyAllowListAddress] by initializing the role of each of
+// the addresses in [AllowListAdmins].
+func (c *AllowListConfig) Configure(state StateDB) {
+	for _, adminAddr := range c.AllowListAdmins {
+		SetAllowListRole(state, adminAddr, AllowListAdmin)
+	}
+}
+
+// Contract returns the singleton stateful precompiled contract to be used for the allow list.
+func (c *AllowListConfig) Contract() StatefulPrecompiledContract { return AllowListPrecompile }
+
+// Valid returns true iff [s] represents a valid role.
+func (s AllowListRole) Valid() bool {
+	switch s {
+	case AllowListNoRole, AllowListEnabled, AllowListAdmin:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsAdmin returns true if [s] indicates the permission to modify the allow list.
+func (s AllowListRole) IsAdmin() bool {
+	switch s {
+	case AllowListAdmin:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsEnabled returns true if [s] indicates that it has permission to access the resource.
+func (s AllowListRole) IsEnabled() bool {
+	switch s {
+	case AllowListAdmin, AllowListEnabled:
+		return true
+	default:
+		return false
+	}
+}
+
+// GetAllowListStatus returns the allow list role of [address].
+func GetAllowListStatus(state StateDB, address common.Address) AllowListRole {
+	// Generate the state key for [address]
+	addressKey := address.Hash()
+	return AllowListRole(state.GetState(AllowListAddress, addressKey))
+}
+
+// SetAllowListRole sets the permissions of [address] to [role]
+// assumes [role] has already been verified as valid.
+func SetAllowListRole(stateDB StateDB, address common.Address, role AllowListRole) {
+	// Generate the state key for [address]
+	addressKey := address.Hash()
+	// Assign [role] to the address
+	stateDB.SetState(AllowListAddress, addressKey, common.Hash(role))
+}
+
+// PackModifyAllowList packs [address] and [role] into the appropriate arguments for modifying the allow list.
+// Note: [role] is not packed in the input value returned, but is instead used as a selector for the function
+// selector that should be encoded in the input.
+func PackModifyAllowList(address common.Address, role AllowListRole) ([]byte, error) {
+	// function selector (4 bytes) + hash for address
+	input := make([]byte, 0, selectorLen+common.HashLength)
+
+	switch role {
+	case AllowListAdmin:
+		input = append(input, setAdminSignature...)
+	case AllowListEnabled:
```

Comment:
nit: AllowListDeployer to be consistent with function sig

Reply:
I actually changed this to `Enabled` so that it is not specific to contract deployments. This way the variable name makes sense if we decide to apply the exact same code to another resource that we want to apply the allow list to, we will just need to do a small refactor to make the address used by these functions an additional parameter and then it will be good to go.

Reply:
If we are doing this, we should change the name of this entire precompile to be `ContractAllowList`? Otherwise, we should keep this internally consistent.